### PR TITLE
more Timeout usage

### PR DESCRIPTION
**Pull request by lihuiba**

### CODE REVIEW

File: net/basic_socket.h [hunk 1]
```diff
@@ -101,36 +105,48 @@ __FORCE_INLINE__ int doio(IOCB iocb, WAIT waitcb) {
...
-    while (v.iovcnt > 0) {
-        ssize_t ret = iocb();
-        if (ret < 0) return ret;
-        if (ret == 0) break;
-        count += ret;
-
-        uint64_t bytes = ret;
-        auto extracted = v.extract_front(bytes);
-        assert(extracted == bytes);
+struct BufAdv {
```
Comment (beef9999) [inline net/basic_socket.h:122]: Adv looks like `add v` but the actual meaning we are going to express is `advance` ?

File: net/basic_socket.h [hunk 2]
```diff
@@ -101,36 +105,48 @@ __FORCE_INLINE__ int doio(IOCB iocb, WAIT waitcb) {
...
+    BufAdv(void*& buf, size_t& count) : buf(buf), count(count) { }
+    BufAdv(size_t& count) : buf(_dummy), count(count) { }
+    bool operator()(size_t ret, size_t n) __INLINE__ {
+        assert(0 <= ret && ret <= count);
+        (char*&)buf += ret;
+        count -= ret;
+        return count > 0;
+    }
+};
+
```
Comment (beef9999) [inline net/basic_socket.h:135]: If V stands for vector

File: net/basic_socket.h (hunk 1 above)
Comment (lihuiba) [reply net/basic_socket.h:122]: Yes, and better naming is welcome.

File: net/basic_socket.h (hunk 2 above)
Comment (lihuiba) [reply net/basic_socket.h:135]: Yes, and better naming is welcome.

**Review by Coldwings — APPROVED** [review]
LGTM

File: net/basic_socket.h (hunk 1 above)
Comment (beef9999) [reply net/basic_socket.h:122]: BufAdvance

Comment (lihuiba) [reply net/basic_socket.h]: how about step?

**Pull request merged**